Docker/validate.py

```diff
@@ -0,0 +1,107 @@
+    """
+    out = {"rownames": df.index.tolist(),
+           "colnames": df.columns.values.tolist()}
+    return out
+
+
+def main():
+    """Main function."""
+    args = get_args()
+
+    invalid_reasons = []
+    gs_file_status = True
+    prediction_file_status = True
+    exp_ids = ["2400", "2401", "7200", "7201"]
+    ds_props = ["0_125"]
+
+    # validate goldstandard file
+    if args.goldstandard is None:
+        gs_file_status = False
+        invalid_reasons = [args.goldstandard + ' not found']
+    else:
+        gs_zip_file = ZipFile(args.goldstandard, "r")
+        true_gs_files = ["pac_" + id + "_gs.csv" for id in exp_ids]
+        gs_diff = list(set(true_gs_files) - set(gs_zip_file.namelist()))
+        # check if all required data exists
+        if gs_diff:
+            invalid_reasons.append("File not found : " + "', '".join(gs_diff))
+            gs_file_status = False
+        else:
+            gs_names = [get_dim_names(gs_f) for gs_f in true_gs_files]
+
+    # validate prediction file
+    if args.submission_file is None:
+        prediction_file_status = False
+        invalid_reasons = [
+            'Expected FileEntity type but found ' + args.entity_type]
+    else:
+        pred_zip_file = ZipFile(args.submission_file, "r")
+        # exp names: pac_{expId}_ds_{prop}_imputed, e.g. pac_1111_ds_0_1
+        true_pred_files = ["pac_" + id + "_ds_" + p +
+                           "_impute.csv" for p in ds_props for id in exp_ids]
+        pred_diff = list(set(true_pred_files) - set(pred_zip_file.namelist()))
+        # check if all required data exists
+        if pred_diff:
+            invalid_reasons.append(
+                "File not found : " + "', '".join(pred_diff))
+            prediction_file_status = False
+        else:
+            true_names = gs_names * len(ds_props)
+            for index, pred_f in enumerate(true_pred_files):
+                pred_df = pd.read_csv(pred_zip_file.open(pred_f), index_col=0)
+                pred_names = get_dim_names(pred_df)
+                true_names = gs_names[index]
+                cp1 = set(pred_names["rownames"]).issubset(
+                    true_names["rownames"])
+                cp2 = set(pred_names["colnames"]).issubset(
+                    true_names["colnames"])
+                # check if names of row/col match with what in goldstandard for each exp
+                if not (cp1 and cp2):
+                    invalid_reasons.append(
+                        pred_f + ": do not contain all genes or cells")
+                    prediction_file_status = False
+                # check if all value is not less than 0
+                elif (pred_df < 0).any().any():
+                    invalid_reasons.append(
+                        pred_f + ": Negative value is not allowed")
+                    prediction_file_status = False
+
+    validate_status = "VALIDATED" if gs_file_status & prediction_file_status else "INVALID"
```

Comment:
Not sure if you already know this, but generally, when doing logical expressions, it's more common to use `and` and `or`, rather than `&` and `|`.

One benefit to the logical operators is that they're [short-circuiting](https://docs.python.org/3/library/stdtypes.html#boolean-operations-and-or-not), that is, if you are using the `and` operator (where both values have to be `True`), Python will immediately stop evaluating once it sees that `gs_file_status` is `False`. Whereas with `&`, Python will continue to evaluate `prediction_file_status`.  Execution time doesn't really matter for this script, since it's short, but in case you ever have to write something more complex in the future :-)

Reply:
An alternative expression you could use is

```python
validate_status = "INVALID" if invalid_reasons else "VALIDATED"
```

where `if invalid_reasons` would return `False` when it's empty (a.k.a. no error messages).  This could help reduce the number of variables in the program.

Reply:
Ah, great suggestions. That is really helpful for me to know python a bit more! Going to change that.